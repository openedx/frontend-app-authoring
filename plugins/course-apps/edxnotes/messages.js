import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  heading: {
    id: 'course-authoring.pages-resources.notes.heading',
    defaultMessage: 'Configure notes',
  },
  enableNotesLabel: {
    id: 'course-authoring.pages-resources.notes.enable-notes.label',
    defaultMessage: 'Notes',
  },
  enableNotesHelp: {
    id: 'course-authoring.pages-resources.notes.enable-notes.help',
    defaultMessage: `Learners can access their notes either in the body of the
    course of on a notes page. On the notes page, a learner can see all the
    notes made during the course. The page also contains links to the location
    of the notes in the course body.`,
  },
  enableNotesLink: {
    id: 'course-authoring.pages-resources.notes.enable-notes.link',
    defaultMessage: 'Learn more about notes',
  },
});

export default messages;
