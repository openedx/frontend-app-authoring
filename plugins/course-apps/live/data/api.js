/* eslint-disable import/prefer-default-export */
import { ensureConfig, getConfig } from '@edx/frontend-platform';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { bbbPlanTypes } from '../constants';

ensureConfig([
  'STUDIO_BASE_URL',
], 'Course Apps API service');

const apiBaseUrl = getConfig().STUDIO_BASE_URL;

export const providersApiUrl = `${apiBaseUrl}/api/course_live/providers`;
export const providerConfigurationApiUrl = `${apiBaseUrl}/api/course_live/course`;

function normalizeProviders(data) {
  const apps = Object.entries(data.providers.available).map(([key, app]) => ({
    id: key,
    featureIds: app.features,
    name: app.name,
    piiSharing: app.pii_sharing,
    hasFreeTier: app.has_free_tier,
  }));

  return {
    activeAppId: data.providers.active,
    selectedAppId: data.providers.active,
    apps,
  };
}

function normalizeLtiConfig(data) {
  if (!data || Object.keys(data).length < 1) {
    return {};
  }

  return {
    consumerKey: data.lti_1p1_client_key,
    consumerSecret: data.lti_1p1_client_secret,
    launchUrl: data.lti_1p1_launch_url,
    launchEmail: data.lti_config?.additional_parameters?.custom_instructor_email,
    tierType: data.tierType,
  };
}

export function normalizeSettings(data) {
  let tier;
  if (data.provider_type === 'big_blue_button') {
    tier = data.free_tier === true ? bbbPlanTypes.free : bbbPlanTypes.commercial;
  }
  return {
    enabled: data.enabled,
    piiSharingAllowed: data.pii_sharing_allowed,
    appConfig: {
      id: data.provider_type,
      ...normalizeLtiConfig({ ...data.lti_configuration, tierType: tier }),
    },
  };
}

export function deNormalizeSettings(data) {
  const ltiConfiguration = {};
  if (data.consumerKey) {
    ltiConfiguration.lti_1p1_client_key = data.consumerKey;
  }
  if (data.consumerSecret) {
    ltiConfiguration.lti_1p1_client_secret = data.consumerSecret;
  }
  if (data.launchUrl) {
    ltiConfiguration.lti_1p1_launch_url = data.launchUrl;
  }
  if (data?.provider === 'zoom' || data.tierType !== 'Free') {
    ltiConfiguration.lti_config = {};
    if (data.launchEmail) {
      ltiConfiguration.lti_config.additional_parameters = {
        custom_instructor_email: data.launchEmail,
      };
    }
  }
  if (Object.keys(ltiConfiguration).length > 0) {
    // Only add this in if we're sending LTI fields.
    // TODO: Eventually support LTI v1.3 here.
    ltiConfiguration.version = 'lti_1p1';
  }

  const apiData = {
    enabled: data?.enabled || false,
    lti_configuration: Object.keys(ltiConfiguration).length ? ltiConfiguration : undefined,
    provider_type: data?.provider || 'zoom',
    pii_sharing_allowed: data?.piiSharingEnable || false,
    free_tier: data?.provider === 'zoom' ? false : Boolean(data.tierType === 'Free'),
  };
  return apiData;
}

/**
 * Fetches providers for provided course
 * @param {string} courseId
 * @returns {Promise<[{}]>}
 */
export async function getLiveProviders(courseId) {
  const { data } = await getAuthenticatedHttpClient()
    .get(`${providersApiUrl}/${courseId}/`);

  return normalizeProviders(data);
}

/**
 * Fetches provider settings for provided course
 * @param {string} courseId
 * @returns {Promise<[{}]>}
 */
export async function getLiveConfiguration(courseId) {
  const { data } = await getAuthenticatedHttpClient()
    .get(`${providerConfigurationApiUrl}/${courseId}/`);
  return normalizeSettings(data);
}

export async function postLiveConfiguration(courseId, config) {
  const { data } = await getAuthenticatedHttpClient().post(
    `${providerConfigurationApiUrl}/${courseId}/`,
    deNormalizeSettings(config),
  );
  return normalizeSettings(data);
}
