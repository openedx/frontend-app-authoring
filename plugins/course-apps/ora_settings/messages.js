import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  heading: {
    id: 'course-authoring.pages-resources.ora.heading',
    defaultMessage: 'Configure Flexible Peer Grading',
    description: 'Title for the modal dialog header',
  },
  ORASettingsHelpLink: {
    id: 'course-authoring.pages-resources.ora.flex-peer-grading.link',
    defaultMessage: 'Learn more about open response assessment settings',
    description: 'Descriptive text for the hyperlink to the docs site',
  },
  enableFlexPeerGradeLabel: {
    id: 'course-authoring.pages-resources.ora.flex-peer-grading.label',
    defaultMessage: 'Flex Peer Grading',
    description: 'Label for form switch',
  },
  enableFlexPeerGradeHelp: {
    id: 'course-authoring.pages-resources.ora.flex-peer-grading.help',
    defaultMessage: 'Turn on Flexible Peer Grading for all open response assessments in the course with peer grading.',
    description: 'Help text describing what happens when the switch is enabled',
  },
  enabledBadgeLabel: {
    id: 'course-authoring.pages-resources.ora.flex-peer-grading.enabled-badge.label',
    defaultMessage: 'Enabled',
    description: 'Label for badge that show users that a setting is enabled',
  },
  cancelLabel: {
    id: 'course-authoring.pages-resources.ora.flex-peer-grading.cancel-button.label',
    defaultMessage: 'Cancel',
    description: 'Label for button that cancels user changes',
  },
  saveLabel: {
    id: 'course-authoring.pages-resources.ora.flex-peer-grading.save-button.label',
    defaultMessage: 'Save',
    description: 'Label for button that saves user changes',
  },
  pendingSaveLabel: {
    id: 'course-authoring.pages-resources.ora.flex-peer-grading.pending-save-button.label',
    defaultMessage: 'Saving',
    description: 'Label for button that has pending api save calls',
  },
  errorSavingTitle: {
    id: 'course-authoring.pages-resources.ora.flex-peer-grading.save-error.title',
    defaultMessage: 'We couldn\'t apply your changes.',
  },
  errorSavingMessage: {
    id: 'course-authoring.pages-resources.ora.flex-peer-grading.save-error.message',
    defaultMessage: 'Please check your entries and try again.',
  },
});

export default messages;
