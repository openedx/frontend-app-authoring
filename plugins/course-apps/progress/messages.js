import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  heading: {
    id: 'course-authoring.pages-resources.progress.heading',
    defaultMessage: 'Configure progress',
  },
  enableProgressLabel: {
    id: 'course-authoring.pages-resources.progress.enable-progress.label',
    defaultMessage: 'Progress',
  },
  enableProgressHelp: {
    id: 'course-authoring.pages-resources.progress.enable-progress.help',
    defaultMessage: `As students work through graded assignments, scores
        will appear under the progress tab. The progress tab contains a chart of
        all graded assignments in the course, with a list of all assignments and
        scores below.`,
  },
  enableProgressLink: {
    id: 'course-authoring.pages-resources.progress.enable-progress.link',
    defaultMessage: 'Learn more about progress',
  },
  enableGraphLabel: {
    id: 'course-authoring.pages-resources.progress.enable-graph.label',
    defaultMessage: 'Enable progress graph',
  },
  enableGraphHelp: {
    id: 'course-authoring.pages-resources.progress.enable-graph.help',
    defaultMessage: 'If enabled, students can view the progress graph',
  },
});

export default messages;
