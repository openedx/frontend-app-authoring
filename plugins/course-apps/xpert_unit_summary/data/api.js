import { getConfig } from '@edx/frontend-platform';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';

export function getXpertSettingsUrl(courseId) {
  return `${getConfig().STUDIO_BASE_URL}/ai_aside/v1/${courseId}`;
}

export function getXpertConfigurationStatusUrl(courseId) {
  return `${getConfig().STUDIO_BASE_URL}/ai_aside/v1/${courseId}/configurable`;
}

export async function getXpertSettings(courseId) {
  const { data } = await getAuthenticatedHttpClient()
    .get(getXpertSettingsUrl(courseId));

  return data;
}

export async function postXpertSettings(courseId, state) {
  const { data } = await getAuthenticatedHttpClient()
    .post(getXpertSettingsUrl(courseId), {
      enabled: state.enabled,
      reset: state.reset,
    });

  return data;
}

export async function getXpertPluginConfigurable(courseId) {
  const { data } = await getAuthenticatedHttpClient()
    .get(getXpertConfigurationStatusUrl(courseId));

  return data;
}

export async function deleteXpertSettings(courseId) {
  const { data } = await getAuthenticatedHttpClient()
    .delete(getXpertSettingsUrl(courseId));

  return data;
}
