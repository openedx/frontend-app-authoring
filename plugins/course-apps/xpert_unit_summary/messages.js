import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  heading: {
    id: 'course-authoring.pages-resources.xpert-unit-summary.heading',
    defaultMessage: 'Configure Xpert unit summaries',
  },
  enableXpertUnitSummaryLabel: {
    id: 'course-authoring.pages-resources.xpert-unit-summary.enable-xpert-unit-summary.label',
    defaultMessage: 'Xpert unit summaries',
  },
  enableXpertUnitSummaryHelp: {
    id: 'course-authoring.pages-resources.xpert-unit-summary.enable-xpert-unit-summary.help',
    defaultMessage: 'Reinforce learning concepts by sharing text-based course content with OpenAI (via API) to display unit summaries on-demand for learners. Learners can leave feedback about the quality of the AI-generated summaries for use by edX to improve the performance of the tool.',
  },
  enableXpertUnitSummaryHelpPrivacyLink: {
    id: 'course-authoring.pages-resources.xpert-unit-summary.enable-xpert-unit-summary.help.privacylink',
    defaultMessage: 'Learn more about OpenAI API data privacy.',
  },
  enableXpertUnitSummaryLink: {
    id: 'course-authoring.pages-resources.xpert-unit-summary.enable-xpert-unit-summary.link',
    defaultMessage: 'Learn more about how OpenAI handles data',
  },
  allUnitsEnabledByDefault: {
    id: 'course-authoring.pages-resources.xpert-unit-summary.all-units-enabled-by-default',
    defaultMessage: 'All units enabled by default',
  },
  noUnitsEnabledByDefault: {
    id: 'course-authoring.pages-resources.xpert-unit-summary.no-units-enabled-by-default',
    defaultMessage: 'No units enabled by default',
  },
});

export default messages;
