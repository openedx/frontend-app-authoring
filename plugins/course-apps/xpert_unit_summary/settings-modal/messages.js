import { defineMessages } from '@edx/frontend-platform/i18n';

const messages = defineMessages({
  cancel: {
    id: 'course-authoring.pages-resources.app-settings-modal.button.cancel',
    defaultMessage: 'Cancel',
  },
  save: {
    id: 'course-authoring.pages-resources.app-settings-modal.button.save',
    defaultMessage: 'Save',
  },
  saving: {
    id: 'course-authoring.pages-resources.app-settings-modal.button.saving',
    defaultMessage: 'Saving',
  },
  saved: {
    id: 'course-authoring.pages-resources.app-settings-modal.button.saved',
    defaultMessage: 'Saved',
  },
  retry: {
    id: 'course-authoring.pages-resources.app-settings-modal.button.retry',
    defaultMessage: 'Retry',
  },
  enabled: {
    id: 'course-authoring.pages-resources.app-settings-modal.badge.enabled',
    defaultMessage: 'Enabled',
  },
  disabled: {
    id: 'course-authoring.pages-resources.app-settings-modal.badge.disabled',
    defaultMessage: 'Disabled',
  },
  resetAllUnits: {
    id: 'course-authoring.pages-resources.app-settings-modal.reset-all-units',
    defaultMessage: 'Reset all units',
  },
  resetAllUnitsTooltipChecked: {
    id: 'course-authoring.pages-resources.app-settings-modal.reset-all-units-tooltip.checked',
    defaultMessage: 'Immediately reset any unit-level changes and checked "Enable summaries" on all units.',
  },
  resetAllUnitsTooltipUnchecked: {
    id: 'course-authoring.pages-resources.app-settings-modal.reset-all-units-tooltip.unchecked',
    defaultMessage: 'Immediately reset any unit-level changes and unchecked "Enable summaries" on all units.',
  },
  reset: {
    id: 'course-authoring.pages-resources.app-settings-modal.reset',
    defaultMessage: 'Reset',
  },
  errorSavingTitle: {
    id: 'course-authoring.pages-resources.app-settings-modal.save-error.title',
    defaultMessage: 'We couldn\'t apply your changes.',
  },
  errorSavingMessage: {
    id: 'course-authoring.pages-resources.app-settings-modal.save-error.message',
    defaultMessage: 'Please check your entries and try again.',
  },
});

export default messages;
