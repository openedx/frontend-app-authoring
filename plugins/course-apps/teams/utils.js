/* eslint-disable import/prefer-default-export */
import { getConfig } from '@edx/frontend-platform';

import { GroupTypes } from 'CourseAuthoring/data/constants';

/**
 * Check if a group type is enabled by the current configuration.
 * This is a temporary workaround to disable the OPEN MANAGED team type until it is fully adopted.
 * For more information, see: https://openedx.atlassian.net/wiki/spaces/COMM/pages/3885760525/Open+Managed+Group+Type
 * @param {string} groupType - the group type to check
 * @returns {boolean} - true if the group type is enabled
 */
export const isGroupTypeEnabled = (groupType) => {
  const enabledTypesByDefault = [
    GroupTypes.OPEN,
    GroupTypes.PUBLIC_MANAGED,
    GroupTypes.PRIVATE_MANAGED,
  ];
  const enabledTypesByConfig = {
    [GroupTypes.OPEN_MANAGED]: getConfig().ENABLE_OPEN_MANAGED_TEAM_TYPE,
  };
  return enabledTypesByDefault.includes(groupType) || enabledTypesByConfig[groupType] || false;
};
