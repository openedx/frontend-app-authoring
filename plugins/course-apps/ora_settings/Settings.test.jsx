import {
  render,
  screen,
  waitFor,
  within,
} from '@testing-library/react';
import ReactDOM from 'react-dom';
import { Routes, Route, MemoryRouter } from 'react-router-dom';
import { initializeMockApp } from '@edx/frontend-platform';
import MockAdapter from 'axios-mock-adapter';
import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
import { AppProvider, PageWrap } from '@edx/frontend-platform/react';
import { IntlProvider } from '@edx/frontend-platform/i18n';

import initializeStore from 'CourseAuthoring/store';
import { executeThunk } from 'CourseAuthoring/utils';
import PagesAndResourcesProvider from 'CourseAuthoring/pages-and-resources/PagesAndResourcesProvider';
import { getCourseAppsApiUrl, getCourseAdvancedSettingsApiUrl } from 'CourseAuthoring/pages-and-resources/data/api';
import { fetchCourseApps, fetchCourseAppSettings } from 'CourseAuthoring/pages-and-resources/data/thunks';
import ORASettings from './Settings';
import messages from './messages';
import {
  courseId,
  inititalState,
} from './factories/mockData';

let axiosMock;
let store;
const oraSettingsUrl = `/course/${courseId}/pages-and-resources/live/settings`;

// Modal creates a portal. Overriding ReactDOM.createPortal allows portals to be tested in jest.
ReactDOM.createPortal = jest.fn(node => node);

const renderComponent = () => (
  render(
    <IntlProvider locale="en">
      <AppProvider store={store} wrapWithRouter={false}>
        <PagesAndResourcesProvider courseId={courseId}>
          <MemoryRouter initialEntries={[oraSettingsUrl]}>
            <Routes>
              <Route path={oraSettingsUrl} element={<PageWrap><ORASettings onClose={jest.fn()} /></PageWrap>} />
            </Routes>
          </MemoryRouter>
        </PagesAndResourcesProvider>
      </AppProvider>
    </IntlProvider>,
  )
);

const mockStore = async ({
  apiStatus,
  enabled,
}) => {
  const settings = ['forceOnFlexiblePeerOpenassessments'];
  const fetchCourseAppsUrl = `${getCourseAppsApiUrl()}/${courseId}`;
  const fetchAdvancedSettingsUrl = `${getCourseAdvancedSettingsApiUrl()}/${courseId}`;

  axiosMock.onGet(fetchCourseAppsUrl).reply(
    200,
    [{
      allowed_operations: { enable: false, configure: true },
      description: 'setting',
      documentation_links: { learnMoreConfiguration: '' },
      enabled,
      id: 'ora_settings',
      name: 'Flexible Peer Grading for ORAs',
    }],
  );
  axiosMock.onGet(fetchAdvancedSettingsUrl).reply(
    apiStatus,
    { force_on_flexible_peer_openassessments: { value: enabled } },
  );

  await executeThunk(fetchCourseApps(courseId), store.dispatch);
  await executeThunk(fetchCourseAppSettings(courseId, settings), store.dispatch);
};

describe('ORASettings', () => {
  beforeEach(async () => {
    initializeMockApp({
      authenticatedUser: {
        userId: 3,
        username: 'abc123',
        administrator: false,
        roles: [],
      },
    });
    store = initializeStore(inititalState);
    axiosMock = new MockAdapter(getAuthenticatedHttpClient());
  });

  it('Flexible peer grading configuration modal is visible', async () => {
    renderComponent();
    expect(screen.getByRole('dialog')).toBeVisible();
  });

  it('Displays "Configure Flexible Peer Grading" heading', async () => {
    renderComponent();
    const headingElement = screen.getByText(messages.heading.defaultMessage);

    expect(headingElement).toBeVisible();
  });

  it('Displays loading component', () => {
    renderComponent();
    const loadingElement = screen.getByRole('status');

    expect(within(loadingElement).getByText('Loading...')).toBeInTheDocument();
  });

  it('Displays Connection Error Alert', async () => {
    await mockStore({ apiStatus: 404, enabled: true });
    renderComponent();
    const errorAlert = screen.getByRole('alert');

    expect(within(errorAlert).getByText('We encountered a technical error when loading this page.', { exact: false })).toBeVisible();
  });

  it('Displays Permissions Error Alert', async () => {
    await mockStore({ apiStatus: 403, enabled: true });
    renderComponent();
    const errorAlert = screen.getByRole('alert');

    expect(within(errorAlert).getByText('You are not authorized to view this page', { exact: false })).toBeVisible();
  });

  it('Displays title, helper text and badge when flexible peer grading button is enabled', async () => {
    renderComponent();
    await mockStore({ apiStatus: 200, enabled: true });

    waitFor(() => {
      const label = screen.getByText(messages.enableFlexPeerGradeLabel.defaultMessage);
      const enableBadge = screen.getByTestId('enable-badge');

      expect(label).toBeVisible();

      expect(enableBadge).toHaveTextContent('Enabled');
    });
  });

  it('Displays title, helper text and hides badge when flexible peer grading button is disabled', async () => {
    renderComponent();
    await mockStore({ apiStatus: 200, enabled: false });

    const label = screen.getByText(messages.enableFlexPeerGradeLabel.defaultMessage);
    const enableBadge = screen.queryByTestId('enable-badge');

    expect(label).toBeVisible();

    expect(enableBadge).toBeNull();
  });
});
