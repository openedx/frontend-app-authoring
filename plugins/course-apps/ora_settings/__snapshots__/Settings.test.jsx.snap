// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ORASettings should render 1`] = `
<AppSettingsModal
  appId="ora_settings"
  hideAppToggle={true}
  initialValues={
    Object {
      "enableFlexiblePeerGrade": "abitrary value",
    }
  }
  onClose={[MockFunction onClose]}
  onSettingsSave={[Function]}
  title={
    <div>
      <p>
        Configure open response assessment
      </p>
      <div
        className="pt-3"
      >
        <withDeprecatedProps(Hyperlink)
          className="text-primary-500 small"
          destination="https://learnmore.test"
          rel="noreferrer noopener"
          target="_blank"
        >
          Learn more about open response assessment settings
        </withDeprecatedProps(Hyperlink)>
      </div>
    </div>
  }
  validationSchema={
    Object {
      "enableFlexiblePeerGrade": "Yub.boolean",
    }
  }
>
  <Component />
</AppSettingsModal>
`;
